api/src/test/resources/calculationsTest.xml

```diff
@@ -0,0 +1,129 @@
+<?xml version='1.0' encoding='UTF-8'?>
```

Comment:
Lets rename this file to be specific to the PVLS calculations, something like `integrationPvlsCalculations.xml`. Better that we isolate the data context for each test set. Will make it more manageable moving forward :+1:

Reply:
it could end up also being used for other non pvls calculations if we will have any! it takes time to setup a test dataset and its why i work of https://github.com/openmrs/openmrs-core/blob/master/api/src/test/resources/org/openmrs/include/standardTestDataset.xml until my scenarios require more data.

Reply:
It is better that we group the necessary datasets, 2 reasons
1. If we need to modify the existing dataset for one test class it could break another test class because the scenarios are different.
2. This single dataset could become very large

In general, I think it is better to maintain vertical slices in how we layout our code. So maybe lets rather have one base dataset to contain the concepts and global properties but keep separate the encounters and obs because those will be unique to the scenarios we are trying to test in each test class.

Make sense?

Reply:
but u still are better off sharing most of those encounters and obs across different scenarios, if we had no idea of reusing things in this dataset and it wasn't still small enough, it would make sense breaking it down. the encounters and obs are all tied around the same concepts we define and so i tend to be attracted to the approach of a grouped test dataset per each calculator after applying the general dataset but in this case these calculators all look tied into one another and it looks ok to me to have one dataset for them to make it easier for people to write new unit tests or even maintain existing ones

Reply:
There is nothing stopping us from using an existing dataset in a new test, especially if the test scenarios are all covered (which means implementing new tests is still quick and easy), however if we tie everything into one dataset then it becomes a problem later on if/when we want to split them into separate files.

Better that we design with this in mind from the beginning.

Remember we want to follow the single responsiblity principle as much as possible and this applies with tests as well. Better to keep each test as isolated as possible (and within reason) from other tests so that the tests can run in a reliable and predictable environment. So using separate dataset files to cater for that particular test class makes each dataset smaller and simpler, which results in it being easier to manage and maintain as our test harness grows.

Reply:
fully agree, i just kind of think in this case where we are testing already closely related calculations that break the single responsibility principle, it's fully sensible to have the same dataset, this can probably just act as the starting dataset for all calculations. something close to the core standardTestDataset.xml.

Reply:
Well I can see problems will come with the encounters and obs for other calculations (perhaps not even needed in other tests) which will mean we will need to either add to this dataset or extend it with another dataset (now breaking single responsibility). Remember the obs and encounters are going to look different depending on the scenarios being tested, therefore there is no guarantee they will be shared across all test classes. 

So my suggestion is to have one base dataset for concepts and global properties that apply to all tests and separate datasets (one for PVLS, one for breastfeeding, etc...) for the encounters and obs which apply to the tests scenarios being covered in those respective classes. There is no guarantee these obs will be used across all test classes therefore they should not be in the shared base dataset

Reply:
am not really implying that your suggestion isn't good @nthfloor, but its just not proper to reason having datasets basing on just encounters and obs, the idea of having many datasets isn't liked to me personally when testing closely related features and considering my 7 years around OpenMRS, those 3 calculators for-example depend on one another except one. having more than 6 dataset files isn't practical and i would not only consider it a less good use of dev time but also making maintenance of code harder; having to look through tens of dataset files before being able to understand closely related tests isn't good design. 
about having one base dataset for just concepts and global properties, i still consider it not perfect since all those we are tracking down for these calculations dataset are for them, loading all of them for other calculators in the base test.
the most practical would be each dataset for each test class since whether concepts or global properties or encounters or obs are closely tied together per each calculation. however it's still not good enough applying 3 datasets for a calculation that runs other calculations. 
THE main reason why i still consider our one dataset more good is it's small and easy to track down things if one is aware that standardTestDataset.xml is our base dataset

Reply:
First off I think that is an over exaggeration that you will need to search through lots of dataset files for each test. What I am proposing that you will need to load 2 maybe 3 datasets (one shared with concepts and one specific with obs and encounters).

Let me ask you something, in the event you pick up a test class and start analysing it, will it be easier to analyse one huge dataset file or a much smaller file which has all the data relevant to that particular test?

This PR https://github.com/esaude/openmrs-module-eptsreports/pull/40/files demonstrates this nicely where the encounters and obs are different, and I think it is a good example of why having separate dataset files is  a good idea since how that test class is very easy to review and maintain.

Maybe others can also comment, @ningosi @psmatsinhe @ynurmahomed @Guimino @edrisse ?

Reply:
obs and encounters require concepts, having a separate file for each set is unpractical,  having all concepts and then just another dataset for obs and encounters, understanding the later is tied on opening and reading through the former, but the most side effect is why load all concepts and gps in a base dataset and then break down obs and encounters into another file.

sure, understanding this datasets with not even 200 lines is easy, it saves time just tracing out matched ids and uuids in separate files.

you see! what's nice to one dev may never be similarly nice to another, am ok with other calculation tests defining extra datasets to add on the base one which exactly that commit does treating the existing one as a base dataset the way it does the standard one in openmrs, but the idea is the base one should have just enough data for shared across calculations.

any way, others could give their view about defineing all concepts and gps in a base dataset and encounters and obs in child datasets!! which would be more easy to read...

Reply:
Make the data set too centralized and you will end up with a massive data set that is hard to understand and maintain; Make it too small and you will have hardwork every time you change the data model. So I guess that something in the middle is the way to go. Normally, on the projects I've worked in the past we use a single data set for each test class and I think that's a good common ground. In our specific case that we are working with openmrs data model, that almost never changes I think we have more headroom to make the data sets even more specific. In general I would prefer more specific data sets as it is easier to maintain and understand.

Reply:
so, if am getting you well @edrisse, in this case u would prefer duplicating the same data if not defining new data when testing RoutineCalculation which runs both the other 2 calculations? i don't like huge datasets too and very much like more specific ones, the idea now is the balance hits at dependent calculations :-)